src/AI/WallyAI.java

```diff
@@ -6,21 +6,46 @@
 import CommandingOfficers.Commander;
 import CommandingOfficers.CommanderAbility;
 import Engine.*;
+import Engine.GamePath.PathNode;
+import Engine.Utils.SearchNode;
 import Engine.Combat.BattleSummary;
 import Engine.Combat.CombatEngine;
+import Engine.Combat.StrikeParams;
 import Engine.Combat.CombatContext.CalcType;
-import Engine.UnitActionLifecycles.BattleLifecycle;
+import Engine.UnitActionLifecycles.CaptureLifecycle;
+import Engine.UnitActionLifecycles.BattleLifecycle.BattleAction;
+import Engine.UnitActionLifecycles.CaptureLifecycle.CaptureAction;
 import Engine.UnitActionLifecycles.WaitLifecycle;
 import Terrain.*;
 import Units.*;
 import Units.MoveTypes.MoveType;
 
-/**
- *  Wally values units based on firepower and the area they can threaten.
- *  He tries to keep units safe by keeping them out of range, but will also meatshield to protect more valuable units.
- */
 public class WallyAI extends ModularAI
 {
+  public WallyAI(Army army)
+  {
+    super(army);
+    aiPhases = new ArrayList<AIModule>(
+        Arrays.asList(
+            new PowerActivator(army, CommanderAbility.PHASE_TURN_START),
+            new DrainActionQueue(army, this),
+            new GenerateThreatMap(army, this), // FreeRealEstate and Travel need this, and NHitKO/building do too because of eviction
+
+            new WallyCapper(army, this),
+            new NHitKO(army, this),
+            new SiegeAttacks(army, this),
+            new PowerActivator(army, CommanderAbility.PHASE_BUY),
+            new BuildStuff(army, this),
+            new FreeRealEstate(army, this),
+            new SiegeTravel(army, this),
+            new Travel(army, this),
+            new Eviction(army, this), // Getting dudes out of the way
+
+            new FillActionQueue(army, this),
+            new PowerActivator(army, CommanderAbility.PHASE_TURN_END)
+            ));
```

Comment:
Why is `FillActionQueue` at the end, and `DrainActionQueue` is at the beginning? That sounds backwards, which means I don't understand the design intent here.

Reply:
Quoth Muriel:
```java
  @Override
  public GameAction getNextAction(GameMap gameMap)
  {
    // If we have actions ready to go, don't bother calculating anything.
    if( !queuedActions.isEmpty() )
    {
      GameAction action = queuedActions.poll();
      log(String.format("  Action: %s", action));
      return action;
    }
```
That's the action queue.
The rest of the `ActionFinders` push actions into an unordered planning map (`mapPlan` IIRC), and `FillActionQueue` is responsible for figuring out the order to *execute* those actions.
So:
I do any queued actions
I build a plan
I turn the plan into an action queue (and return the first action)

Reply:
Oh man. Choosing actions and then choosing the order to execute them? Sounds dicey (though maybe that explains the logic to catch unexpected divergences and clear the rest of the plan).

Reply:
Yessir, dicey it is.
I have discovered a lovely little case where Wally will plan a perfectly viable series of attacks, then reassign one of the guys who was supposed to walk away to create space to dive in and attack, borking everything.
Haven't fixed it yet, though I have made progress.

Reply:
He creates the plan, then instead of executing the plan, he just... changes the plan? That's a strange way to use a plan. Good luck though, I guess. These changes are getting pretty in the weeds.

Reply:
He considers a re-plan whenever he's planned to kill off a new unit, because that opens up new tiles for new attacks.
But I wasn't properly accounting for the order the actions would have to be in, which would break things.
With [769b3ee](https://github.com/ThislsAUsername/DefendPeace/pull/430/commits/769b3ee3a9d4b74e78640b4a7fdfc3a132a0b2f2), I managed to fix that such that (if you disabled the travel calc steps), he could plan and execute moves like this:
(Wally is blue, yellow is an ally)
![image](https://github.com/ThislsAUsername/DefendPeace/assets/8674241/fd9d159b-66ad-43ea-9450-6b1c7a184f97)
![image](https://github.com/ThislsAUsername/DefendPeace/assets/8674241/ae67da2d-a471-4f71-849c-fabfedb16746)

But as I mentioned, having the other compute steps on messed it up.
I tried to address that, then got distracted with other stuff.

I've been tinkering with this on and off for ages, so sorry it's kind of all over the place.

Reply:
The tests I've added should give some insight into the sorts of things I'm trying to do, as well.
Was the "good luck" an indication that I can merge this whenever I decide it's ready? hah.

Reply:
> I can merge this whenever I decide it's ready?

More or less, yeah. Wally is your thing, and I'm not gonna get too involved with changes to his code, plus there is a _lot_ going on here, with a good mix of "yeah, makes sense" and "how does this even run?". I've at least looked at all the changes at this point, but I'm trying to do more of a skim than a dive.

Reply:
Understandable.
I am fairly interested in examples of the latter.

Reply:
I mean it's more "this doesn't make immediate sense, and I don't feel like spending half an hour figuring out how it does make sense, and if it didn't work you would probably have fixed it so it did".

Like, `canReachHitFromZone` - name doesn't really tell me what it's doing, and the first thing it does is grab the starting/root node and check if it has a parent; if not, it returns immediately. Root nodes are typically orphans (cry), so at first blush it seems like the rest of the function is unreachable.

I'm doubtful it would be worth complaining every time I didn't immediately understand what was going on, especially in Wally-specific code.

Reply:
Ah, yeah.

"HitFromZone" is supposed to indicate "the places this particular threat can hit this particular tile from".
Each one of them is a reachable tile for said threat, assuming there are no units in the way. (and the parent of each node is the tile that you would path through to get to it)
The function tells you if all of the HitFromZone is blocked (by either having units on or blocking access to those tiles).

Probably fair.
It's also probably fair to say I should add more documentation, but the function name itself is usually enough for me.

Reply:
> the function name itself is usually enough for me.

And that's the difference between coding for yourself, and coding for the homicidal maniac who knows where you live and will be maintaining your code after you.

Or in my case, "future me who forgot everything I've written more than a year ago". Maybe you don't have that problem, but I sometimes do, and especially when I'm reading code written by not-me.

Reply:
Said homicidal maniac should know that this house of cards is necessarily interdependent and stateful, so making almost any change is at least a little fraught.
Like, there's a flag that controls whether Wally's units should fear death or not. Turning it off generally leads to better swarming/inf wall tactics, but also means his artillery think kissing enemy tanks is a good hobby.
I've tried a few strategies for toggling it, but each has had substantial failings.
We have decent test coverage for "Can you plan your way out of this paper bag?", but building good unit tests for complex tactical scenarios sounds hard.

For a slightly better example, I think I've gotten the "plan a string of actions that will be executable in a clogged map state" thing set up to the point where most of the state it needs is "on stack", and it can generate a sequence of actions that makes sense.
But trying to make it generic would add complexity I don't need, and would probably rule out some of the "other state" interactions it has to deconflict with already-planned action sequences.
Maybe that's a good argument for more documentation, not less.

I do forget sometimes as well, but my style of naming things is a pretty high-density language in my head, heh.